src/utils.jl

```diff
@@ -1167,3 +1167,54 @@ _argument_description(s::Symbol) =
 
 _document_argument(s::Symbol) =
     _fmt_paragraph(_argument_description(s), leadingspaces = 6 + length(string(s)))
+
+#
+# The following functions implement the guess of the optimal legend position,
+# from the data series.
+#
+# Computes the distances of the plot limits to a sample of points at the extremes of 
+# the ranges, and places the legend at the corner where the maximum distance to the limits is found.
+#
+function _dmin_point(dmin, x, y, lim, scale)
+    p_scaled = (x / scale[1], y / scale[2])
+    d = sum(abs2, lim .- p_scaled)
+    dmin = ifelse(d < dmin, d, dmin) # preferred relative to min because x < NaN === false 
+    return dmin
+end
+function _dmin_series(lim, scale, x, y, nsamples)
+    length(x) > 0 || return +Inf
+    dmin = +Inf
+    step = 1
+    lim = lim ./ scale
+    n = max(1, div(min(nsamples, length(x)), 2))
+    # Run from the extremes of the dataset inwards
+    for isample in firstindex(x):step:n
+        dmin = _dmin_point(dmin, x[isample], y[isample], lim, scale)
+    end
+    for isample in lastindex(x):(-step):(lastindex(x) - n)
+        dmin = _dmin_point(dmin, x[isample], y[isample], lim, scale)
+    end
+    return dmin
+end
+#
+# Function barrier because lims are type-unstable
+#
+function _guess_best_legend_position(xl, yl, plt, nsamples)
+    scale = (maximum(xl) - minimum(xl), maximum(yl) - minimum(yl))
+    dist_to_lims = fill(+Inf, 4) # faster than tuple
+    for series in plt.series_list
+        x = series[:x]
+        y = series[:y]
+        for (i, lim) in enumerate(Iterators.product(xl, yl))
+            dist_to_lims[i] = min(dist_to_lims[i], _dmin_series(lim, scale, x, y, nsamples))
+        end
+    end
+    # this inversion favors :topright in case of draws, without cost
+    ibest = findmax(@view(dist_to_lims[4:-1:1]))[2]
```

Comment:
Couldn't we detect overlap here by adding a distance threshold ?
Such that we use `findfirst` instead and on failure (`nothing`) return `:topright` as fallback.

Reply:
To actually detect overlap we need the coordinates of the legend box. With that information we could compute at roughly the same cost how many atoms fall into it. 

Using any other heuristic without that  may be of course fine in particular cases, but I'm not sure if it is worth the trouble.

Reply:
I dislike introducing heuristics in a deterministic algorithm too.
I'm happy with the current, but I'm also anticipating users complaining about the new behavior.
Let me think about that while ci is running.

Reply:
I didn't mean something with aleatory behavior, just a different strategy.

I'm pretty sure someone will complain, but that will happen with any strategy :-D

What I can try rather easily is to use a sum of inverse distances, or similar, and see if you like the results better.

Reply:
Regarding the `coordinates of the legend box`, the legend geometry is computed here https://github.com/JuliaPlots/Plots.jl/blob/816d0f4c5cda98f13300cdf265338995d58723e5/src/backends/gr.jl#L1248-L1265

And you namedtuple `leg` is accessible in the parent calling function https://github.com/JuliaPlots/Plots.jl/blob/816d0f4c5cda98f13300cdf265338995d58723e5/src/backends/gr.jl#L1105, so you can maybe pass this nt to `_guess_best_legend_position` (and make it conditional, defaulting to `nothing` for other backends that might not compute legend geometry).

My little drawing in https://github.com/JuliaPlots/Plots.jl/blob/816d0f4c5cda98f13300cdf265338995d58723e5/src/backends/gr.jl#L1219-L1224 might be useful here.

Reply:
I'm against calling backend functions in frontend code. We'd need a utility function in the frontend then, that backends can overload.